### ISSUE TITLE
ES modules

### ISSUE THREAD

**Issue by hdgarrood** [purescript/purescript#3613]
Previously: #2207, #2558, #2574, #2575, #3427. I'm making a new issue because the discussions in many of the linked issues are quite long, and because the ES modules landscape has changed quite a lot since when most of the discussion happened two years ago (in particular, all major browsers now support ES modules). Also, we should probably implement emitting ES modules at the same time as switching to ES modules for FFI, so I think all of these things should be considered together.

ES modules provide a number of benefits over CommonJS; most notably easier dead code elimination and wide browser support (see <https://caniuse.com/#feat=es6-module>). Having the compiler use and understand ES modules natively would mean that many workflows wouldn't require bundling at all, which would significantly speed up and simplify things, especially during development.

It should remain possible to run a PureScript-only app in modern browsers and in Node.js with just `purs`, i.e. without having to bring in external JS tools such as webpack. Currently this is achieved by invoking `purs compile` and then `purs bundle` (or just `purs compile` if you only want to run on Node.js).

I think the following is relatively uncontroversial, and is what we generally have settled on in the above linked issues:
* Compile import declarations in PureScript to ES module import statements, so that `import Prelude` compiles to something like `import * as Prelude from "../Prelude/index.js"`. Note that `import Prelude` currently compiles to `var Prelude = require("../Prelude/index.js")`.
* Compile exports as ES module exports, perhaps `export { apply, identity }` rather than the current `module.exports = { apply: apply, identity: identity }`
* Mangle primes in imports and exports in both regular PureScript modules and in FFI modules, so for example if a PureScript module exports an identifier `foo'`, then the compiled ES module exports it as `foo$prime`. In particular, note that writing `foreign import foo' :: Int` in a PureScript file means that the compiler should look for `foo$prime` in the corresponding FFI module.
* Support writing FFI modules both as ES modules and with the current CommonJS format, at least for a period of time.

There are some slightly harder questions to be resolved about how we approach FFI files still. The only part of the compiler which needs to care about the format of FFI modules is `purs bundle` of course, but we should also consider other workflows (i.e. bundling with tools like rollup or webpack, or running directly in Node.js). For example: what file extension should we use? I'd prefer to avoid `.mjs` and `.cjs` if we possibly can, as they're only a temporary Node.js thing, and lots of existing tools and scripts will expect to find things under e.g. `output/*/*.js`. In fact I think if we were using `.js` files everywhere, including for FFI modules (which could be a mixture of ES modules and CommonJS), the only setup where we might have problems would be apps running in Node.js with no bundling step, and perhaps that's not too much of a problem: if you're using Node.js there's a good chance you are already using webpack or babel or something similar anyway, but if not it's probably not too difficult to set one up, either with one of the JS bundling tools or with `purs bundle`.

I don't think we should have people use the `.mjs` extension for FFI files which are written as ES modules; instead, I'd prefer to always use the `.js` extension and always parse FFI modules in strict mode / module mode, as I expect the proportion of FFI modules which would parse differently in strict mode is very small and possibly even zero (most include `"use strict";` directives already). We can then handle the syntax trees of both CommonJS modules and ES modules after parsing.

**Comment by hdgarrood** [comment]
I should probably also address the question of whether the compiler should be able to produce both output modes (ESM and CommonJS) via a flag. I'd much prefer to not do that: maintaining both ESM and CommonJS output at once would be a lot more difficult, and most workflows should be able to handle ES module output from the compiler without much adjustment, and possibly without any adjustment at all, since the plan is that `purs bundle` will still produce the same format.

**Comment by kevinbarabash** [comment]
I'm excited about this.  Support for ES6 modules in node is still experimental but it does support the .js extension when the nearest package.json has `"type": "module"` set, see https://nodejs.org/api/esm.html#esm_enabling.

In addition to mangling primes we should probably mangle keywords and `null` as well.

**Comment by hdgarrood** [comment]
> In addition to mangling primes we should probably mangle keywords and null as well.

That's not necessary and I actually think it would be better not to; see discussion in #2558.

**Comment by gasi** [comment]
@twitchard and I spent this weekend’s [NYC CoHack Haskell hackathon](https://www.meetup.com/NY-Haskell/events/261245898/) setting up a test bench for evaluating dead code elimination (DCE) / tree shaking using webpack + PureScript. Please give it a shot: https://github.com/gasi/purescript-dce-benchmark#purescript-dead-code-elimination-proof-of-concept

We used basic TDD to see status quo and where my [fork of the compiler](https://github.com/gasi/purescript/pull/3) gets us. 

As many of the referenced issues point out: We are currently blocked by the lack of support for ES2015+ in `language-javascript`. In https://github.com/gasi/purescript/pull/2, I propose we replace `language-javascript` with the full-fledged `tree-sitter` parser and its backing by GItHub.

Curious to hear other people’s thoughts. Would love to move the PureScript compiler + ecosystem to ES modules 😄

**Comment by kl0tl** [comment]
I’d like to take a stab at this issue now that `language-javascript` can parse ES modules.

According to this [benchmark](https://github.com/kl0tl/purescript-es-modules-benchmark) of my ongoing fork it would allow Webpack 5 and Rollup to outperform the optimizations done by `purs bundle` and `zephyr`!

**Comment by hdgarrood** [comment]
That would be great! If you start with @gasi’s code, I think the main task is getting `purs bundle` to work with both ESM and CommonJS. Now that I think of it, there are also questions to resolve about how we approach the FFI primes issue so as to minimise breakage. I would be tempted to do it in two stages: first, spend one release cycle (eg all of 0.14.x) warning if anyone is using primes in FFI modules in a way which would become invalid, and then in the next breaking release after that, start generating ESM, and drop support for primes in FFI modules. It would be great if we could have eg `foreign import foo' :: Int` work for JS modules which export either `foo'` or `foo$prime` during the first phase of this process so that people can update in preparation for the change, eg by having `foreign import foo' :: Int` compile to `var foo$prime = $foreign[$foreign.hasOwnProperty("foo'") ? "foo'" : "foo$prime"]` or something along those lines.

**Comment by kl0tl** [comment]
I’ve opened https://github.com/purescript/purescript/pull/3791! I’ll work on the dual support of escaped and unescaped primes in identifiers exported from foreign modules in another PR.

**Comment by natefaubion** [comment]
Maybe we should go ahead and add a warning in the next release, deprecating foreign primes?

**Comment by hdgarrood** [comment]
So we'll warn on this throughout 0.14, and aim to get ESM done for 0.15? Sounds good to me :)

**Comment by shmish111** [comment]
I'm really looking forward to this since more and more libraries are set up in a way that makes it fiddly to use in PS because you can't import them directly in the FFI js files. Plus of course the more modern optimization in bundling tools.

**Comment by realvictorprm** [comment]
Well so far I see it's not possible to emit and consume template string literals due to this right now :(

**Comment by hdgarrood** [comment]
Yes, it’s not possible. However, template strings are unrelated to this issue. We have no plans to support template strings right now, although it may already be possible to sensibly achieve what you want to do; please direct questions to Slack or Discourse.

**Comment by mhmdanas** [comment]
@realvictorprm  This issue is for ES modules. Plenty of other ES6 features are already supported, like property value shorthands, `const`, and `let`. Template literals might be supported, but I'm not sure. You can open an issue at `language-javascript` for template literals.

**Comment by artemisSystem** [comment]
Should this issue be closed now that PureScript 0.15 is released?
